### PR TITLE
gcp-observability: remove logging channel/server providers

### PR DESCRIPTION
**Pull request by DNVindhya**
This PR removes `ManagedChannelProvider` and  `ServerProvider` used by logging.
 
With this change, going forward all three features of Observability namely logging, metrics and traces will use `GlobalInterceptors` to set Client/Server interceptors as well as Stream Tracers.

CC @sanjaypujare

### CODE REVIEW

File: gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java [hunk 1]
```diff
@@ -97,13 +97,8 @@ static GcpObservability grpcInit(
...
     if (instance == null) {
-      instance =
-          new GcpObservability(sink, config, channelInterceptorFactory, serverInterceptorFactory);
-      LogHelper logHelper = new LogHelper(sink, TimeProvider.SYSTEM_TIME_PROVIDER);
-      ConfigFilterHelper logFilterHelper = ConfigFilterHelper.factory(config);
-      instance.setProducer(
-          new InternalLoggingChannelInterceptor.FactoryImpl(logHelper, logFilterHelper),
-          new InternalLoggingServerInterceptor.FactoryImpl(logHelper, logFilterHelper));
+      instance = new GcpObservability(sink, config);
+      instance.setProducer(channelInterceptorFactory, serverInterceptorFactory);
```
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java:101]: Something that's now clear is that we only ever create one instance of the `InternalLoggingChannelInterceptor` and `InternalLoggingServerInterceptor` each and install those in the Global interceptors. So the factory is kind of useless. One thing we may consider (may be a separate PR) is to eliminate the Factories and replace those with those respective interceptors.

File: gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java [hunk 2]
```diff
@@ -160,6 +154,7 @@ private void registerStackDriverExporter(String projectId, Map<String, String> c
                 e -> LabelValue.create(e.getValue())));
         statsConfigurationBuilder.setConstantLabels(constantLabels);
       }
+      statsConfigurationBuilder.setExportInterval(Duration.create(METRICS_EXPORT_INTERVAL, 0));
```
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java:157]: Instead of hardcoding, should we make this configurable? Is there agreement with other languages about the export interval to StackDriver to have a hard coded interval of 30 seconds?

File: gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/InternalLoggingChannelInterceptor.java [hunk 3]
```diff
@@ -40,7 +40,7 @@
 import java.util.logging.Logger;
 
 /**
- * A logging interceptor for {@code LoggingChannelProvider}.
+ * A logging client interceptor for Observability.
```
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/InternalLoggingChannelInterceptor.java:43]: As mentioned elsewhere we should consider getting rid of the Factory (in the Server interceptor as well). A separate PR seems more convenient?

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java [hunk 4]
```diff
@@ -98,8 +83,9 @@ public GcpLogSink(String destinationProjectId, Map<String, String> locationTags,
   @Override
   public void write(GrpcLogRecord logProto) {
     if (gcpLoggingClient == null) {
-      logger.log(Level.SEVERE, "Attempt to write after GcpLogSink is closed.");
-      return;
+      synchronized (this) {
+        gcpLoggingClient = createLoggingClient();
```
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:87]: Don't you need to check again if `gcpLoggingClient == null`  since the first check was outside the lock?

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java [hunk 5]
```diff
@@ -133,6 +119,14 @@ public void write(GrpcLogRecord logProto) {
     }
   }
 
+  Logging createLoggingClient() {
```
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:122]: nit: the location of this method changed which seems unnecessary. Or did it change because of the access modifier?

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java [hunk 6]
```diff
@@ -54,36 +54,21 @@ public class GcpLogSink implements Sink {
       = ImmutableSet.of("project_id", "location", "cluster_name", "namespace_name",
       "pod_name", "container_name");
   private static final long FALLBACK_FLUSH_LIMIT = 100L;
+  private final String projectId;
   private final Map<String, String> customTags;
-  private final Logging gcpLoggingClient;
   private final MonitoredResource kubernetesResource;
   private final Long flushLimit;
+  private Logging gcpLoggingClient;
```
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:61]: nit: it will be good to add a comment about why we are doing lazy initialization of `gcpLoggingClient` .

File: gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java (hunk 2 above)
Comment (DNVindhya) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java:157]: I think making it configurable gives more flexibility to user, which is definitely favorable IMO. 
We did not have a discussion on using a specific value, since Golang [implementation](https://github.com/lidizheng/grpc-go/blob/master/gcp/observability/opencensus.go#L38) set it to 30 seconds, I did the same.

File: gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/InternalLoggingChannelInterceptor.java (hunk 3 above)
Comment (DNVindhya) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/InternalLoggingChannelInterceptor.java:43]: I do agree and will do this in another PR.

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java (hunk 5 above)
Comment (DNVindhya) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:122]: I modified it in order to keep public methods before package private and private methods. I also notice that in order to achieve that  `close()` needs to be in put in top of the class along with other public methods.
If you think this is not necessary, I will undo this change.

File: gcp-observability/src/test/java/io/grpc/gcp/observability/GcpObservabilityTest.java [hunk 7]
```diff
@@ -85,8 +83,6 @@ public static final class StaticTestingClassInitFinish implements Runnable {
     @Override
     public void run() {
       // TODO(dnvindhya) : Remove usage of Providers on cleaning up Logging*Provider
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/GcpObservabilityTest.java:85]: Remove this TODO?

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java (hunk 5 above)
Comment (sanjaypujare) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:122]: No it's okay, not a big deal.

File: gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java (hunk 2 above)
Comment (sanjaypujare) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java:157]: It seems okay for now but I think this should have been flagged earlier and added to the o11y config as a configurable value. Do we even know why 30 seconds is a reasonable value? If you don't call `setExportInterval` what is the default used by the `statsConfigurationBuilder` ?

May be create an internal bug to address by the config design owner.

Comment (DNVindhya) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java:157]: Default export interval for `StackdriverStatsConfiguration` is [60 seconds](https://github.com/census-instrumentation/opencensus-java/blob/master/exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsConfiguration.java#L47).  From local testing, what i observed was if the application is completed before that duration the metrics were not exported. 
Also, filed an internal bug

File: gcp-observability/src/test/java/io/grpc/gcp/observability/MetricsTest.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  private Map<String, String> customTags = Collections.singletonMap(customTagKey, customTagValue);
+
+  @Rule public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+
+  /**
+   * Cloud Monitoring test using GlobalInterceptors.
+   *
+   * <p>Ignoring test, because it calls external Cloud Monitoring APIs. To test cloud monitoring
+   * setup locally, 1. Set up Cloud auth credentials 2. Assign permissions to service account to
+   * write metrics to project specified by variable PROJECT_ID 3. Comment @Ignore annotation
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/MetricsTest.java:65]: I am assuming when enabled with the correct set up this test passes? It might be worth mentioning that.

File: gcp-observability/src/test/java/io/grpc/gcp/observability/MetricsTest.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    ListTimeSeriesRequest metricsRequest =
+        ListTimeSeriesRequest.newBuilder()
+            .setName(ProjectName.of(PROJECT_ID).toString())
+            .setFilter(metricsFilter)
+            .setInterval(interval)
+            .build();
+    ListTimeSeriesPagedResponse response = metricServiceClient.listTimeSeries(metricsRequest);
+    assertThat(response.iterateAll()).isNotEmpty();
+    for (TimeSeries ts : response.iterateAll()) {
+      assertThat(ts.getPoints(0).getValue().getInt64Value()).isEqualTo(1);
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/MetricsTest.java:124]: Can we also assert the method-name value (as one of the attached tags)?

File: gcp-observability/src/test/java/io/grpc/gcp/observability/TracesTest.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+  @Rule public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+
+  /**
+   * Cloud Trace test using GlobalInterceptors.
+   *
+   * <p>Ignoring test, because it calls external Cloud Tracing APIs. To test cloud trace setup
+   * locally, 1. Set up Cloud auth credentials 2. Assign permissions to service account to write
+   * traces to project specified by variable PROJECT_ID 3. Comment @Ignore annotation
+   */
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/TracesTest.java:69]: Similar to the other comments, it might be worth mentioning that the test has been verified manually

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java [hunk 11]
```diff
@@ -98,32 +100,31 @@ public class GcpLogSinkTest {
       .putFields("rpc_id", Value.newBuilder().setStringValue(rpcId).build())
       .build();
   private Logging mockLogging;
+  private GcpLogSink spySink;
 
   @Before
   public void setUp() {
     mockLogging = mock(Logging.class);
+    spySink = spy(new GcpLogSink(destProjectName, locationTags,
+        customTags, flushLimit));
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:109]: This code brings up a couple of points:

- it looks a little weird that `GcpLogSink` is the class-under-test and that class itself is being mocked or spied. Usually all the other external classes are mocked.

- overloading `createLoggingClient()` in a subclass would have been an option then I was wondering why `GcpLogSink` is not final. Should it not be (even if it makes testing difficult)?

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java [hunk 12]
```diff
@@ -98,32 +100,31 @@ public class GcpLogSinkTest {
...
+    Mockito.doReturn(mockLogging).when(spySink).createLoggingClient();
   }
 
   @Test
   public void createSink() {
-    Sink mockSink = new GcpLogSink(mockLogging, destProjectName, locationTags,
-        customTags, flushLimit);
-    assertThat(mockSink).isInstanceOf(GcpLogSink.class);
+    assertThat(spySink).isInstanceOf(GcpLogSink.class);
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:115]: what does this test achieve? Or what does it really test from the "class-under-test"?

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java [hunk 13]
```diff
@@ -98,32 +100,31 @@ public class GcpLogSinkTest {
...
-    mockSink.write(logProto);
+    spySink.write(logProto);
 
     ArgumentCaptor<Collection<LogEntry>> logEntrySetCaptor = ArgumentCaptor.forClass(
         (Class) Collection.class);
     verify(mockLogging, times(1)).write(logEntrySetCaptor.capture());
     for (Iterator<LogEntry> it = logEntrySetCaptor.getValue().iterator(); it.hasNext(); ) {
       LogEntry entry = it.next();
-      System.out.println(entry);
       assertThat(entry.getPayload().getData()).isEqualTo(expectedStructLogProto);
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:128]: Just noticed `expectedStructLogProto` - can it be defined in upper-case as `EXPECTED_STRUCT_LOG_PROTO` since it is like a constant object?

Comment (sanjaypujare) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:128]: similarly `logProto` ?

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java [hunk 14]
```diff
@@ -98,32 +100,31 @@ public class GcpLogSinkTest {
       .putFields("rpc_id", Value.newBuilder().setStringValue(rpcId).build())
       .build();
   private Logging mockLogging;
+  private GcpLogSink spySink;
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:103]: Only some tests use `spySink` - others use a local `spySink2`  which looks a little odd. May be we can get rid of this class member and just use a local `spySink` in all tests based on the test's requirements?

**Review by sanjaypujare — COMMENTED** [review]
finished 1st round. Will do the 2nd round once the comments are addressed

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java (hunk 4 above)
Comment (DNVindhya) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:87]: Done

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java (hunk 6 above)
Comment (DNVindhya) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:61]: Added. PTAL

File: gcp-observability/src/test/java/io/grpc/gcp/observability/GcpObservabilityTest.java (hunk 7 above)
Comment (DNVindhya) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/GcpObservabilityTest.java:85]: Done

File: gcp-observability/src/test/java/io/grpc/gcp/observability/MetricsTest.java (hunk 9 above)
Comment (DNVindhya) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/MetricsTest.java:124]: Done

File: gcp-observability/src/test/java/io/grpc/gcp/observability/TracesTest.java (hunk 10 above)
Comment (DNVindhya) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/TracesTest.java:69]: Added. PTAL

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java (hunk 11 above)
Comment (DNVindhya) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:109]: Updated to use test only constructor

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java (hunk 13 above)
Comment (DNVindhya) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java]: Updated

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java (hunk 14 above)
Comment (DNVindhya) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java]: Done

**Comment by DNVindhya** [comment]
Addressed comments. PTAL

File: gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java (hunk 1 above)
Comment (sanjaypujare) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/GcpObservability.java:101]: How about a TODO comment saying we will eliminate the factory?

File: gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/InternalLoggingChannelInterceptor.java (hunk 3 above)
Comment (sanjaypujare) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/InternalLoggingChannelInterceptor.java:43]: Pls add a TODO comment saying so

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java [hunk 15]
```diff
@@ -97,34 +96,29 @@ public class GcpLogSinkTest {
...
-  }
-
   @Test
   public void createSink() {
-    Sink mockSink = new GcpLogSink(mockLogging, destProjectName, locationTags,
-        customTags, flushLimit);
-    assertThat(mockSink).isInstanceOf(GcpLogSink.class);
+    GcpLogSink sink = new GcpLogSink(mockLogging, destProjectName, locationTags,
+        customTags, FLUSH_LIMIT);
+    assertThat(sink).isInstanceOf(Sink.class);
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:106]: What's the value of this assert? Since `GcpLogSink` is declared to implement `Sink` this will be true so the test seems trivial - does not verify any run time behavior

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java [hunk 16]
```diff
@@ -64,18 +63,18 @@ public class GcpLogSinkTest {
...
-  private final String rpcId = "d155e885-9587-4e77-81f7-3aa5a443d47f";
-  private final GrpcLogRecord logProto = GrpcLogRecord.newBuilder()
+  private static final long seqId = 1;
+  private static final String destProjectName = "PROJECT";
+  private static final String serviceName = "service";
+  private static final String methodName = "method";
+  private static final String authority = "authority";
+  private static final Duration timeout = Durations.fromMillis(1234);
+  private static final String rpcId = "d155e885-9587-4e77-81f7-3aa5a443d47f";
+  private static final GrpcLogRecord LOG_PROTO = GrpcLogRecord.newBuilder()
```
Comment (sanjaypujare) [inline gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:77]: All of the above are constants so why not follow the coding convention and declare in all upper-case? From `locationTags` to `rpcId`

**Review by sanjaypujare — CHANGES_REQUESTED** [review]
Almost there. Small changes requested

File: gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/InternalLoggingChannelInterceptor.java (hunk 3 above)
Comment (DNVindhya) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/interceptors/InternalLoggingChannelInterceptor.java:43]: Done

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java (hunk 15 above)
Comment (DNVindhya) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java]: Makes sense. Removed the test since this was not adding any additional value.

File: gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java (hunk 16 above)
Comment (DNVindhya) [reply gcp-observability/src/test/java/io/grpc/gcp/observability/logging/GcpLogSinkTest.java:79]: Updated.

**Review by sanjaypujare — APPROVED** [review]
LGTM

**Comment by sanjaypujare** [comment]
CC @ejona86
